internal/command/command_service.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package command
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/google/uuid"
+	mpi "github.com/nginx/agent/v3/api/grpc/mpi/v1"
+	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/grpc"
+	"github.com/nginx/agent/v3/internal/logger"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	backoffHelpers "github.com/nginx/agent/v3/internal/backoff"
+)
+
+var _ commandService = (*CommandService)(nil)
+
+const (
+	retryInterval                  = 5 * time.Second
+	createConnectionMaxElapsedTime = 0
+)
+
+type CommandService struct {
+	commandServiceClient mpi.CommandServiceClient
+	agentConfig          *config.Config
+	isConnected          *atomic.Bool
+	subscribeCancel      context.CancelFunc
+	subscribeMutex       sync.Mutex
+}
+
+func NewCommandService(
+	commandServiceClient mpi.CommandServiceClient,
+	agentConfig *config.Config,
+) *CommandService {
+	isConnected := &atomic.Bool{}
+	isConnected.Store(false)
+
+	return &CommandService{
+		commandServiceClient: commandServiceClient,
+		agentConfig:          agentConfig,
+		isConnected:          isConnected,
+	}
+}
+
+func (cs *CommandService) UpdateDataPlaneStatus(ctx context.Context, resource *mpi.Resource) error {
+	if !cs.isConnected.Load() {
+		err := cs.createConnection(ctx, resource)
+		if err != nil {
+			return err
+		}
+
+		var subscribeCtx context.Context
+
+		cs.subscribeMutex.Lock()
+		subscribeCtx, cs.subscribeCancel = context.WithCancel(ctx)
```

Comment:
Are we sure we want to pass the parent context to `context.WithCancel` here? I would think that we wouldn't want to close the gRPC connection if the context of a Data Plane Status request cancelled 🤔

I'd have thought we only cancel the `subscribe` goroutine when `CancelSubscription` is called?

Reply:
Moved the subscribe function to the creation of the command service instead.
Also added a subscribe channel for the management plane requests.